write('get.product.sequences.R', p("jobs/", opt$sequence.name, "/pipeline.checkpoint.txt"))
#FOR SANGER SEQUENCING. 
#Reads primer3 output, then splits the products into smaller sequences for sanger sequencing.
#Then generates primer3 input files to find primers for these smaller sequences.

gene.name = opt$sequence.name
project.path = base_directory

grab.primer3.output.value = function(output.file.path, property.id){
  #grab the value of a specified property of a primer 3 output file.
  #input: primer3 output file as read by readLines()
  #args:
  # output.file.path - character string of the path to the primer3 output file
  # property.id - character string of the property of interest, e.g. "PRIMER_LEFT_0"
  
  current.file.lines = readLines(output.file.path)
  prop.coord = grep(paste(property.id, "=", sep = ""), current.file.lines)
  if(length(prop.coord) == 0){
    return("Property not found")
  } else {
    pen1 = current.file.lines[prop.coord]
    pen1 = strsplit(pen1, "=")
    return(pen1[[1]][2])
  }
}

output.files = list.files(p(project.path, "jobs/", gene.name, "/primers/best.primers/set1/"))

print('p(project.path, "jobs/", gene.name, "/primers/best.primers/set1/")')
print(p(project.path, "jobs/", gene.name, "/primers/best.primers/set1/"))


print("output.files")
print(output.files)


#parse the user input template sequence from primer3 output files and cut it into the products generated by the various primer sets
products = lapply(output.files, function(x){
  current.file = p(project.path, "jobs/", gene.name, "/primers/best.primers/set1/", x)
  
  template.sequence = DNAString(grab.primer3.output.value(current.file, "SEQUENCE_TEMPLATE"))
  
  left.primer.pos = grab.primer3.output.value(current.file, "PRIMER_LEFT_0")
  left.primer.pos = strsplit(left.primer.pos, ",")
  left.primer.pos = unlist(lapply(left.primer.pos, as.numeric))
  left.primer.start = left.primer.pos[1] + 1
  
  
  right.primer.pos = grab.primer3.output.value(current.file, "PRIMER_RIGHT_0")
  right.primer.pos = strsplit(right.primer.pos, ",")
  right.primer.pos = unlist(lapply(right.primer.pos, as.numeric))
  right.primer.end = right.primer.pos[1] + 1
  
  template.sequence[left.primer.start:right.primer.end]
})

product.ids = strsplit(output.files, "\\.")
product.ids = unlist(lapply(product.ids, function(x){
  x[2]
}))

#
generate.ranges = function(product.size){
  #makes a dataframe of overlapping ranges for cutting up a DNA sequence
  #args:
  # product.size: integer, the size of the DNA sequence to cut in base pairs
  # num.divisions = ceiling(product.size / 500)
  if(product.size < 501){
    start.ranges = seq(1, product.size, (product.size / 1.5))  
  } else {
    start.ranges = seq(1, product.size, 500)  
  }
  
  end.ranges = c((start.ranges - 1)[-1], product.size)
  
  #add some buffer to the start ranges 
  start.ranges.extended = start.ranges
  start.ranges.extended[2:length(start.ranges.extended)] = start.ranges.extended[2:length(start.ranges.extended)] - 120
  
  all.ranges = data.frame(start.ranges.extended, end.ranges)
  all.ranges
}

#cut the products into overlapping ranges as defined by generate.ranges()
products.split = lapply(products, function(x){
  ranges = generate.ranges(length(x))
  products.split = list()
  for(i in 2:nrow(ranges)){
    products.split = c(products.split, x[ranges[i, 1]:ranges[i, 2]])
  }
  products.split
})

names(products.split) = product.ids

generate.primer3.input.files = function(template.sequence2, p3.seqid, product.size.min, product.size.max){
  #args:
  # template.sequence2 - a DNAString object without inserts ("-"s)
  # p3.seqid - character string indicating name of sequence (used both inside the primer3 input file and in the title of the primer3 input file)
  
  
  
  #primer3 variables:
  p3.template = as.character(template.sequence2)
  p3.product.size.range = paste(as.character(product.size.min), "-", as.character(product.size.max), sep = "")
  
  #note here that line breaks "\n" have to be added in manually as 
  #writeLines automatically adds a line break to the end of every line,
  #whilst primer3_core will not accept a file in which the last line 
  #has a line break on it
  primer3.input = c(p("SEQUENCE_ID=", p3.seqid, "\n"), 
                    p("SEQUENCE_TEMPLATE=", p3.template, "\n"),
                    p("PRIMER_PRODUCT_SIZE_RANGE=", p3.product.size.range, "\n"),
                    "=")
  
  output.filepath = file(p(project.path, "jobs/", gene.name, "/primers/sequencing.primers/primer3.", p3.seqid, ".txt"), "wb")
  writeLines(primer3.input, output.filepath, sep = "")
  close(output.filepath)
  "Done"
}

count1 = make.counter()
namecounter = make.counter()
lapply(products.split, function(x){
  product.name = product.ids[namecounter()]
  lapply(x, function(y){
    # browser()
    generate.primer3.input.files(template.sequence2 = as.character(y), p3.seqid = paste(product.name, ".", count1(), sep = ""), product.size.min = (length(y) - 80), product.size.max = length(y))
  })
})

primer3inputfiles = list.files(p(project.path, "jobs/", gene.name, "/primers/sequencing.primers/"), pattern = ".txt")

#run primer3 to generate primers
if(file.exists(p(project.path, "jobs/", gene.name, "/primers/sequencing.primers/input"))) {
  NA
} else {
  dir.create(p(project.path, "jobs/", gene.name, "/primers/sequencing.primers/input"))
}

if(file.exists(p(project.path, "jobs/", gene.name, "/primers/sequencing.primers/output"))) {
  NA
} else {
  dir.create(p(project.path, "jobs/", gene.name, "/primers/sequencing.primers/output"))
}

lapply(primer3inputfiles, function(x){
  file.copy(p(project.path, "jobs/", gene.name, "/primers/sequencing.primers/", x), p(project.path, "jobs/", gene.name, "/primers/sequencing.primers/input/", x))
  file.remove(p(project.path, "jobs/", gene.name, "/primers/sequencing.primers/", x))
  #browser()
  system(p("primer3_core < ", project.path, "jobs/", gene.name, "/primers/sequencing.primers/input/", x, " > ", project.path, "jobs/", gene.name, "/primers/sequencing.primers/output/", x, ".output.txt"))

})

